### PR TITLE
Split swagger UI and rework auth to support routes that require access from both TRE Admins and Workspace Users

### PR DESCRIPTION
**Pull request by marrobi**
# PR for issue #1062 

## What is being addressed

Split swagger UI and rework auth to support routes that require access from both TRE Admins and Workspace Users.

Graph calls and Directory.ReadAll  isnow only needed  if a user needs to list their workspaces.

## How is this addressed
I've changed more than I wanted to. However does all seem to work.

Swagger for TRE Admin/TRE User at: /api/docs
Swagger for Workspace Owner, Workspace Researcher as /api/workspaces/(specific_workspace_id)/docs

### CODE REVIEW

File: api_app/services/aad_authentication.py [hunk 1]
```diff
@@ -1,60 +1,75 @@
...
-            tokenUrl=f"{aad_instance}/{aad_tenant}/oauth2/v2.0/token",
-            refreshUrl=f"{aad_instance}/{aad_tenant}/oauth2/v2.0/token",
-            scheme_name="oauth2", scopes={
-                f"api://{config.API_CLIENT_ID}/Workspace.Read": "List and Get TRE Workspaces",
-                f"api://{config.API_CLIENT_ID}/Workspace.Write": "Modify TRE Workspaces"
-            },
+            authorizationUrl=f"{config.AAD_INSTANCE}/{config.AAD_TENANT_ID}/oauth2/v2.0/authorize",
+            tokenUrl=f"{config.AAD_INSTANCE}/{config.AAD_TENANT_ID}/oauth2/v2.0/token",
+            refreshUrl=f"{config.AAD_INSTANCE}/{config.AAD_TENANT_ID}/oauth2/v2.0/token",
+            scheme_name="oauth2",
```
Comment (marrobi) [inline api_app/services/aad_authentication.py:31]: Should revisit scopes - do we need to verify scopes? The scopes need to be different if its a workspace operation vs TRE API operation.

File: api_app/services/aad_authentication.py [hunk 2]
```diff
@@ -128,7 +144,94 @@ def _get_token_key(self, key_id: str) -> str:
 
         return AzureADAuthorization._jwt_keys[key_id]
 
+    # The below functions are needed to list which workspaces a specific user has access to i.e. GET /workspaces.
```
Comment (marrobi) [inline api_app/services/aad_authentication.py:147]: Need to add this to the docs

File: api_app/api/routes/workspaces.py [hunk 3]
```diff
@@ -229,3 +251,56 @@ async def patch_user_resource(user_resource_patch: UserResourcePatchEnabled, use
...
+
+workspace_router = APIRouter(prefix=config.API_PREFIX)
+workspace_router.include_router(workspaces_router, tags=["workspaces"])
+workspace_router.include_router(workspace_services_router, tags=["workspace services"])
+workspace_router.include_router(user_resources_router, tags=["user resources"])
+
+workspace_openapi: DefaultDict[str, Optional[Dict[str, Any]]] = defaultdict(lambda: None)
+
+
+@workspace_swagger_router.get("/workspaces/{workspace_id}/openapi.json", include_in_schema=False, name="workspace_openapi")
```
Comment (marrobi) [inline api_app/api/routes/workspaces.py:264]: Need to add new swagger config to the docs.

File: api_app/api/routes/api.py [hunk 4]
```diff
@@ -14,12 +14,15 @@
...
+core_router = APIRouter(prefix=config.API_PREFIX)
+core_router.include_router(health.router, tags=["health"])
+core_router.include_router(workspace_templates.router, tags=["workspace templates"])
+core_router.include_router(workspace_service_templates.workspace_service_templates_router, tags=["workspace service templates"])
+core_router.include_router(workspace_service_templates.user_resource_templates_router, tags=["user resource templates"])
+core_router.include_router(status.router, tags=["status"])
+core_router.include_router(workspaces.tre_router, tags=["workspaces"])
+
+router.include_router(core_router)
+router.include_router(workspaces.workspace_router)
```
Comment (eladiw) [inline api_app/api/routes/api.py:28]: do you also need the 'tag=["workspace"]'?

Comment (eladiw) [reply api_app/api/routes/api.py:28]: is it on purpose that the user resources router got deleted?

File: api_app/api/routes/workspace_service_templates.py [hunk 5]
```diff
@@ -11,27 +11,27 @@
...
 
-service_templates_router = APIRouter(dependencies=[Depends(get_current_tre_user)])
-user_resource_templates_router = APIRouter(dependencies=[Depends(get_current_tre_user)])
+workspace_service_templates_router = APIRouter(dependencies=[Depends(get_current_workspace_owner_or_researcher_user_or_tre_admin)])
+user_resource_templates_router = APIRouter(dependencies=[Depends(get_current_workspace_owner_or_researcher_user_or_tre_admin)])
 
 
 # WORKSPACE SERVICE TEMPLATES
-@service_templates_router.get("/workspace-service-templates", response_model=ResourceTemplateInformationInList, name=strings.API_GET_WORKSPACE_SERVICE_TEMPLATES, dependencies=[Depends(get_current_tre_user)])
+@workspace_service_templates_router.get("/workspace-service-templates", response_model=ResourceTemplateInformationInList, name=strings.API_GET_WORKSPACE_SERVICE_TEMPLATES, dependencies=[Depends(get_current_workspace_owner_or_researcher_user_or_tre_admin)])
```
Comment (eladiw) [inline api_app/api/routes/workspace_service_templates.py:22]: get_current_tre_user became get_current_workspace_owner_or_researcher_user_or_tre_admin.
is that correct. isn't get_current_workspace_owner_or_researcher_user_or_tre_admin equals get_current_ws_user?

Comment (eladiw) [reply api_app/api/routes/workspace_service_templates.py:22]: same for rest of file

File: api_app/services/aad_authentication.py [hunk 6]
```diff
@@ -1,60 +1,75 @@
 import base64
 import logging
+from typing import List
 import jwt
 import requests
 import rsa
 
 from fastapi import Request, HTTPException, status
-from fastapi.security import OAuth2AuthorizationCodeBearer
+from msal import ConfidentialClientApplication
```
Comment (eladiw) [inline api_app/services/aad_authentication.py:9]: why needed?

File: api_app/tests_ma/test_api/test_routes/test_workspace_templates.py [hunk 7]
```diff
@@ -40,7 +39,8 @@ class TestWorkspaceTemplate:
 
     @pytest.fixture(autouse=True, scope='class')
     def _prepare(self, app, admin_user):
-        app.dependency_overrides[get_current_tre_user] = admin_user
+        app.dependency_overrides[get_current_admin_user] = admin_user
+        app.dependency_overrides[get_current_workspace_owner_or_researcher_user_or_tre_admin] = admin_user
```
Comment (eladiw) [inline api_app/tests_ma/test_api/test_routes/test_workspace_templates.py:43]: is it correct that both L42 and L43 are 'adnin_user' ?

**Review by eladiw — CHANGES_REQUESTED** [review]
Some questions

File: api_app/services/aad_authentication.py (hunk 6 above)
Comment (marrobi) [reply api_app/services/aad_authentication.py:9]: Line 153. For connecting to Graph. I moved all the AzureAD stuff into this class.

File: api_app/api/routes/workspace_service_templates.py (hunk 5 above)
Comment (marrobi) [reply api_app/api/routes/workspace_service_templates.py:22]: `get_current_tre_user = AzureADAuthorization(require_one_of_roles=['TREUser'])`

whereas

`get_current_workspace_owner_or_researcher_user_or_tre_admin = AzureADAuthorization(require_one_of_roles=["TREAdmin", "WorkspaceOwner", "WorkspaceResearcher"])`

As some routes need to be accessed by both TREAdmins and any Workspace User. 

 get_current_workspace_owner_or_researcher_user_or_tre_admin does not equal get_current_ws_user as get_current_ws_user would not include TREAdmin. I renamed to be more explicit, so have:

 ```
 get_current_workspace_owner_user = AzureADAuthorization(require_one_of_roles=['WorkspaceOwner'])


get_current_workspace_researcher_user = AzureADAuthorization(require_one_of_roles=['WorkspaceResearcher'])


get_current_workspace_owner_or_researcher_user = AzureADAuthorization(require_one_of_roles=['WorkspaceOwner', 'WorkspaceResearcher'])
```

Let me know if want to talk through.

File: api_app/api/routes/api.py (hunk 4 above)
Comment (marrobi) [reply api_app/api/routes/api.py:28]: Its part of workspaces.workspace_router

the tags are in workspaces.py, maybe we can rename/move things around as is a bit confusing. The worksapces.py has too much in and needs to be split IMHO. But didn't want to chnge too many files.

File: api_app/services/aad_authentication.py (hunk 1 above)
Comment (marrobi) [reply api_app/services/aad_authentication.py:31]: Lets look at integrating scoeps and moving to https://intility.github.io/fastapi-azure-auth/ as a seperate issue.

Comment (marrobi) [reply api_app/services/aad_authentication.py:34]: https://github.com/microsoft/AzureTRE/issues/1074

File: api_app/services/aad_authentication.py [hunk 8]
```diff
@@ -1,60 +1,75 @@
...
         token: str = await super(AzureADAuthorization, self).__call__(request)
 
-        try:
-            app_reg_id = self._default_app_reg_id
+        decoded_token = None
 
-            # if an app reg id was not given, it means we need to fetch it, based on the given workspace id
-            if self._default_app_reg_id is None:
-                logging.info("Default workspace app registration was not provided. Translating from workspace ID")
+        if 'workspace_id' in request.path_params:
```
[the remark this answers is not in the document]
Comment (marrobi) [reply api_app/services/aad_authentication.py]: Added role conditions.

File: api_app/tests_ma/test_api/test_routes/test_workspace_templates.py (hunk 7 above)
Comment (marrobi) [reply api_app/tests_ma/test_api/test_routes/test_workspace_templates.py:43]: That second override was redundant, removed.

File: api_app/api/routes/api.py (hunk 4 above)
Comment (marrobi) [reply api_app/api/routes/api.py]: I've updated the tags, and reorganized the routers, does this look correct?

image

Comment (eladiw) [reply api_app/api/routes/api.py]: looks good

File: api_app/api/routes/workspace_service_templates.py (hunk 5 above)
Comment (eladiw) [reply api_app/api/routes/workspace_service_templates.py]: talked offline. seems correct

File: docs/tre-admins/setup-instructions/installing-workspace-service-and-user-resource.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+1. Paste the following payload json into the `Request body` field, then click `Execute`. Review the server response.
+
+```json
+{
+  "templateName": "tre-service-guacamole",
+  "properties": {
+    "display_name": "Virtual Desktops",
+    "description": "Create virtual desktops for runnign research workloads"
+  }
+}
```
Comment (LizaShak) [inline docs/tre-admins/setup-instructions/installing-workspace-service-and-user-resource.md:74]: Maybe it is worth adding the whole payload needed for the service, something like this:
```suggestion
{
  "templateName": "tre-service-guacamole",
  "properties": {
    "display_name": "Virtual Desktops",
    "description": "Create virtual desktops for runnign research workloads",
    "tre_id": "<TRE ID>",
    "api_client_id": "<TRE API app registration client id>",
    "aad_tenant_id": "<Tenant id>",
    "openid_client_id": "<Workspace app registration client id>"
  }
}
```

Comment (marrobi) [reply docs/tre-admins/setup-instructions/installing-workspace-service-and-user-resource.md]: Updated.

**Review by eladiw — APPROVED** [review]

**Pull request merged**